### PR TITLE
feat: new web-extension package with e2e cip30 test

### PR DESCRIPTION
**Pull request by mkazlauskas**
# Context

Existing cip30 package can't be easily integrated in a web extension-based wallet

# Proposed Solution

- Add a new `web-extension` package that has e2e tests for dapp connector
  - Create the package, set up:
    - [x] [webdriverio](https://webdriver.io/)
    - [x] Load a [test dapp](https://github.com/dynamicstrategies/cardano-wallet-connector) (commited static pre-built files)
    - Encapsulate the complexities of web extension messaging, providing easy to use functions:
      - [x] `exposePromiseApi`: share an object that has promise-based API with other processes
      - [x] `consumeRemotePromiseApi`: use an object that lives in a different extension process
      - [x] `runContentScriptMessageProxy` function that passes through certain messages from web app to background process (to be run in content script).
    - Create and load a test extension
      - [x] Build with webpack
      - Set up the extension using `cip30` and `wallet` packages to build sign a simple transaction from the dapp
        - [x] initialization code (great example for reviewing this PR, under `packages/web-extension/e2e/extension`)
        - [x] basic e2e spec
- Refactor cip30 to support a web extension-based wallet
    - [x] Create a `PersistentAuthenticator` class, that's responsible of maintaining the whitelist
    - [x] Set up AuthenticatorApi to be shared using messaging API, running in background process
    - [x] Set up WalletApi to be shared using messaging API, running in background process
    - [x] Refactor `cip30` `Wallet` (renamed to `Cip30Wallet`) by utilizing an `Authenticator` to replace some it's existing functionality
    - [x] update tests and cover new functionality
- [x] add cip30 getCollateral function (not implemented for SingleAddressWallet - just returns null for now)

See [this diagram](https://drive.google.com/file/d/18-6tIRgXU4Nr2eNvLs4mezIx7ACQwUqJ/view) for an overview of cip30.

# Important Changes Introduced

- [x] add `node_modules` and `dist` to `.eslintignore`
- [x] chore(core): add missing eslint-disable in serializableObject.ts
- [x] make cip30 `Wallet.apiVersion` hardcoded, because version bump will require code change anyway, and only the code knows what it's compatible with 
- [x] refactor(wallet)!: rm obsolete cip30.initialize
- [x] add cslToCore.utxo

## Follow-up work (out of scope of this PR)

- resolve blake2b-wasm load issue and replace stub wallet with the real wallet
- update the test dapp to latest cip30 spec:
  - doesn't support addresses returned as bech32)
  - maybe more issues to be found
- additional test coverage:
  - improve web-extension e2e tests to cover more functionality
  - maybe write more isolated unit or integration tests for web-extension/src/messaging
- some QA for production readiness:
  - find if `runtime.connect` and `port.postMessage` could possibly throw
  - find how long it takes for port.onDisconnect to be called when the other end doesn't exist (worst case scenario)
  - find if we need to do additional validation on message source inside the injected script - it could potentially be sent by a content script of another extension
- implement messaging utils for sharing observable wallet throughout extension
- we'll need to refactor http clients to not use got, because it doesn't support browser env
- refactor e2e tests config to be shared for both `wallet` and `web-extension` packages

### CODE REVIEW

File: packages/cip30/test/Authenticator/PersistentAuthenticator.test.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  });
+
+  describe('revokeAccess', () => {
+    beforeEach(async () => {
+      requestAccess.mockResolvedValueOnce(true);
+      await authenticator.requestAccess(origin1);
+      storage.set.mockReset();
+    });
+
+    it('non-whitelisted origin => returns false', async () => {
```
Comment (rhyslbw) [inline packages/cip30/test/Authenticator/PersistentAuthenticator.test.ts:69]:
```suggestion
    it('unknown origin => returns false', async () => {
```

File: packages/cip30/test/Authenticator/PersistentAuthenticator.test.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      expect(await authenticator.revokeAccess(origin2)).toBe(false);
+    });
+
+    describe('whitelisted origin', () => {
```
Comment (rhyslbw) [inline packages/cip30/test/Authenticator/PersistentAuthenticator.test.ts:73]:
```suggestion
    describe('allowed origin', () => {
```

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/cip30/src/Authenticator/PersistentAuthenticator.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  async requestAccess(origin: Origin) {
+    if (this.#origins.includes(origin)) {
+      return true;
+    }
+    try {
+      const accessGranted = await this.#requestAccess(origin);
+      if (accessGranted) {
+        const newOrigins = [...this.#origins, origin];
+        if (await this.#store(newOrigins)) {
+          this.#logger.info('[Authenticator] whitelisted', origin);
```
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/cip30/src/Authenticator/PersistentAuthenticator.ts]: Renamed all occurrences in [7f08443](https://github.com/input-output-hk/cardano-js-sdk/pull/232/commits/7f084436270c5c5e81db42838769f6b9951de28d)

File: packages/core/src/util/misc/serializableObject.ts [hunk 4]
```diff
@@ -1,9 +1,9 @@
+/* eslint-disable sonarjs/cognitive-complexity */
```
Comment (rhyslbw) [inline packages/core/src/util/misc/serializableObject.ts:1]: The commit message: 
**chore(core): add missing eslint-disable in serializableObject.ts**
... would be better described as:
**chore(core): hoist eslint-disable to whole file**

File: packages/web-extension/e2e/extension/config.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
+// Review: this is copied from packages/wallet
+// and updated to not use ogmios, ledger key agent, txsubmithttpprovider, waitOn, bunyan
+// Because they didn't work in extension out of the box.
+// Ideally we'd reuse this in both packages, util-dev? Only problem is circular deps...
+// Also: we'll need to refactor http clients to not use got, because it doesn't support browser env
```
Comment (rhyslbw) [inline packages/web-extension/e2e/extension/config.ts:5]: :+1:

File: packages/web-extension/e2e/extension/config.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
+// Review: this is copied from packages/wallet
+// and updated to not use ogmios, ledger key agent, txsubmithttpprovider, waitOn, bunyan
+// Because they didn't work in extension out of the box.
+// Ideally we'd reuse this in both packages, util-dev? Only problem is circular deps...
```
Comment (rhyslbw) [inline packages/web-extension/e2e/extension/config.ts:4]: :+1:

File: packages/web-extension/e2e/extension/config.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
+// Review: this is copied from packages/wallet
```
Comment (rhyslbw) [inline packages/web-extension/e2e/extension/config.ts:1]: let's capture it as a separate task. It's fine for now given it's in the test scope

File: packages/web-extension/src/messaging/BackgroundMessenger.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  ports: Set<MessengerPort>;
+  hasMethodRequestHandler?: boolean;
+}
+
+/**
+ * Intended to be used in service worker background process.
+ * Manages connections with different parts of the extension.
+ * Connections are managed through ports.
+ * All other parts of extension are expected use NonBackgroundMessenger.
+ * You won't be able to add any additional 'runtime.onConnect' listeners in background process once this is called.
```
Comment (rhyslbw) [inline packages/web-extension/src/messaging/BackgroundMessenger.ts:17]: Why so?

File: packages/cip30/src/WalletApi/types.ts [hunk 9]
```diff
@@ -54,7 +54,7 @@ export type GetNetworkId = () => Promise<Cardano.NetworkId>;
  * @throws ApiError
  * @throws PaginateError
  */
-export type GetUtxos = (amount?: Cbor, paginate?: Paginate) => Promise<Cardano.Utxo[] | undefined>;
+export type GetUtxos = (amount?: Cbor, paginate?: Paginate) => Promise<Cbor[] | undefined>;
```
Comment (rhyslbw) [inline packages/cip30/src/WalletApi/types.ts:57]: Good catch

File: packages/cip30/src/WalletApi/types.ts [hunk 10]
```diff
@@ -24,7 +54,16 @@ export type GetNetworkId = () => Promise<Cardano.NetworkId>;
...
+
+/**
+ * @returns a list of one or more UTXOs (unspent transaction outputs) controlled by the wallet
+ * that are required to reach AT LEAST the combined ADA value target specified in amount
+ * AND the best suitable to be used as collateral inputs
+ * for transactions with plutus script inputs (pure ADA-only utxos).
```
Comment (rhyslbw) [inline packages/cip30/src/WalletApi/types.ts:63]:
```suggestion
 * for transactions with plutus script inputs (pure ADA-only UTxOs).
```

File: packages/cip30/src/WalletApi/types.ts [hunk 11]
```diff
@@ -24,7 +54,16 @@ export type GetNetworkId = () => Promise<Cardano.NetworkId>;
  * @throws ApiError
  * @throws PaginateError
  */
-export type GetUtxos = (amount?: Cbor, paginate?: Paginate) => Promise<Cardano.Utxo[] | undefined>;
+export type GetUtxos = (amount?: Cbor, paginate?: Paginate) => Promise<Cbor[] | undefined>;
+
+/**
+ * @returns a list of one or more UTXOs (unspent transaction outputs) controlled by the wallet
```
Comment (rhyslbw) [inline packages/cip30/src/WalletApi/types.ts:60]:
```suggestion
 * @returns a list of one or more UTxOs (unspent transaction outputs) controlled by the wallet
```

**Review by rhyslbw — CHANGES_REQUESTED** [review]
Just a few minor things

File: packages/web-extension/src/messaging/BackgroundMessenger.ts (hunk 8 above)
Comment (mkazlauskas) [reply packages/web-extension/src/messaging/BackgroundMessenger.ts:17]: It's something I found via trial and error. Multiple handlers in the same process just won't work.

File: packages/web-extension/e2e/extension/config.ts (hunk 7 above)
Comment (mkazlauskas) [reply packages/web-extension/e2e/extension/config.ts:1]: captured in PR description

File: packages/web-extension/e2e/extension/config.ts (hunk 5 above)
Comment (mkazlauskas) [reply packages/web-extension/e2e/extension/config.ts:5]: captured in PR description

**Review by rhyslbw — APPROVED** [review]
Excellent work @mkazlauskas :tada:

**Pull request merged**